snooty/diagnostics.py

```diff
@@ -0,0 +1,204 @@
+import enum
+from typing import Tuple, Union
+from pathlib import Path
+
+
+class Diagnostic:
+    def __init__(
+        self,
+        message: str,
+        start: Union[int, Tuple[int, int]],
+        end: Union[None, int, Tuple[int, int]] = None,
+    ) -> None:
+        self.message = message
+
+        if isinstance(start, int):
+            start_line, start_column = start, 0
+        else:
+            start_line, start_column = start
+        self.start = (start_line, start_column)
+
+        if end is None:
+            end_line, end_column = start_line, 1000
+        elif isinstance(end, int):
+            end_line, end_column = end, 1000
+        else:
+            end_line, end_column = end
+        self.end = (end_line, end_column)
+
+    class Level(enum.IntEnum):
+        info = 0
+        error = 1
+        warning = 2
+
+        @classmethod
+        def from_docutils(cls, docutils_level: int) -> "Diagnostic.Level":
+            level = docutils_level - 1
+            level = min(level, cls.warning)
+            level = max(level, cls.info)
+            return cls(level)
+
+    @property
+    def severity(self) -> "Diagnostic.Level":
+        raise TypeError("Cannot access the severity of an abstract base Diagnostic")
+
+    @property
+    def severity_string(self) -> str:
+        return self.severity.name.title()
+
+
+class UnexpectedIndentation(Diagnostic):
+    severity = Diagnostic.Level.error
+
+
+class InvalidURL(Diagnostic):
+    severity = Diagnostic.Level.error
+
+
+# this could captuer 'literal include expected path arg..as well?
+class ExpectedPathArg(Diagnostic):
+    severity = Diagnostic.Level.error
+
+    def __init__(
+        self,
+        name: Path,
+        start: Union[int, Tuple[int, int]],
+        end: Union[None, int, Tuple[int, int]] = None,
+    ) -> None:
+        super().__init__(f'"{name}" expected a path argument', start, end)
+        self.name = name
+
+
+class ExpectedImgArg(Diagnostic):
+    severity = Diagnostic.Level.error
+
+
+class ImgExpectedButNotRequired(Diagnostic):
```

Comment:
[nit] Just because this class name is a bit long, do you think `ImgSuggested` would make sense? I think either way is fine though!

Reply:
Shorter is good! Perhaps `ImageSuggested`?

Sophie, you wrote the logic where this is raised originally, right? If it really is just a suggestion, should it be info rather than warning?

Reply:
Yup, info sounds good! And if we go with `ImageSuggested`, we should probably change `ExpectedImgArg` to `ExpectedImageArg` for consistency.

Reply:
Ah good catch